renku/core/metadata/repository.py

```diff
@@ -518,10 +579,69 @@ def get_global_configuration(writable: bool = False) -> "Configuration":
         return Configuration(repository=None, writable=writable)
 
     @staticmethod
-    def hash_object(path: Union[Path, str]) -> str:
+    def list_paths_in_revision(
+        paths: List[Union[Path, str]] = None, revision: str = "HEAD", working_dir="."
+    ) -> List[str]:
+        """List all paths that exist in a revision.
+
+        Warning: ls-tree is sensitive to the current working directory, make sure to set working_dir to repo root.
+        """
+        dirs = []
+        files = []
+
+        for path in paths:
+            if os.path.isdir(path):
+                dirs.append(path)
+            else:
+                files.append(path)
+
+        if dirs or files:
+            result = []
+            try:
+                if files:
+                    for batch in split_paths(*files):
+                        existing_paths = git.Git(working_dir=working_dir).ls_tree(*batch, r=revision, name_only=True)
+                        result.extend(existing_paths.splitlines())
+
+                if dirs:
+                    for batch in split_paths(*dirs):
+                        existing_paths = git.Git(working_dir=working_dir).ls_tree(
+                            *batch, d=True, r=revision, name_only=True
+                        )
+                        result.extend(existing_paths.splitlines())
+
+                return result
+            except git.GitCommandError as e:
+                raise errors.GitCommandError(
+                    message=f"Git command failed: {str(e)}",
+                    command=e.command,
+                    stdout=e.stdout,
+                    stderr=e.stderr,
+                    status=e.status,
+                ) from e
+        else:
+            try:
+                existing_files = git.Git().ls_tree(r=revision, name_only=True).splitlines()
+                existing_dirs = git.Git().ls_tree(r=revision, name_only=True, d=True).splitlines()
+                return existing_dirs + existing_files
+            except git.GitCommandError as e:
+                raise errors.GitCommandError(
+                    message=f"Git command failed: {str(e)}",
+                    command=e.command,
+                    stdout=e.stdout,
+                    stderr=e.stderr,
+                    status=e.status,
+                ) from e
```

Comment:
We should use inner functions to remove redundant calls to ls_tree. and its error handling.

Reply:
I don't think there are redundant calls, there are two calls each for with/without paths, showing directories and showing files, so 4 distinct calls. Moving that to an inner function doesn't really make it more readable.

But the error handling can just be wrapped around the whole function body.